rai/node/bootstrap.cpp

```diff
@@ -950,12 +960,78 @@ bool rai::bootstrap_attempt::consume_future (std::future<bool> & future_a)
 
 void rai::bootstrap_attempt::process_fork (MDB_txn * transaction_a, std::shared_ptr<rai::block> block_a)
 {
+	std::unique_lock<std::mutex> lock (mutex);
+	try_resolve_fork (transaction_a, block_a, true);
+}
+
+void rai::bootstrap_attempt::try_resolve_fork (MDB_txn * transaction_a, std::shared_ptr<rai::block> block_a, bool from_processor)
+{
+	assert (!mutex.try_lock ());
+	std::weak_ptr<rai::bootstrap_attempt> this_w (shared_from_this ());
 	std::shared_ptr<rai::block> ledger_block (node->ledger.forked_block (transaction_a, *block_a));
-	if (!node->active.start (transaction_a, ledger_block))
+	if (ledger_block)
 	{
-		node->network.broadcast_confirm_req (ledger_block);
-		node->network.broadcast_confirm_req (block_a);
-		BOOST_LOG (node->log) << boost::str (boost::format ("While bootstrappping, fork between our block %1% and block %2% both with root %3%") % ledger_block->hash ().to_string () % block_a->hash ().to_string () % block_a->root ().to_string ());
+		node->active.start (transaction_a, ledger_block, [this_w, block_a](std::shared_ptr<rai::block>, bool resolved) {
+			if (auto this_l = this_w.lock ())
+			{
+				if (resolved)
+				{
+					{
+						std::unique_lock<std::mutex> lock (this_l->mutex);
+						this_l->unresolved_forks.erase (block_a->hash ());
+						this_l->condition.notify_all ();
+					}
+					rai::transaction transaction (this_l->node->store.environment, nullptr, false);
+					auto account (this_l->node->ledger.store.frontier_get (transaction, block_a->root ()));
+					if (!account.is_zero ())
+					{
+						this_l->requeue_pull (rai::pull_info (account, block_a->root (), block_a->root ()));
+					}
+					else if (this_l->node->ledger.store.account_exists (transaction, block_a->root ()))
+					{
+						this_l->requeue_pull (rai::pull_info (block_a->root (), rai::block_hash (0), rai::block_hash (0)));
+					}
+				}
+			}
+		});
+
+		auto hash = block_a->hash ();
+		bool exists = unresolved_forks.find (hash) != unresolved_forks.end ();
+		if (!exists)
+		{
+			BOOST_LOG (node->log) << boost::str (boost::format ("While bootstrappping, fork between our block: %1% and block %2% both with root %3%") % ledger_block->hash ().to_string () % hash.to_string () % block_a->root ().to_string ());
+			unresolved_forks[hash] = block_a;
+		}
+		if (!exists || !from_processor)
+		{
+			// Only broadcast if it's a new fork, or if the request is coming from the retry loop.
+			node->network.broadcast_confirm_req (ledger_block);
+			node->network.broadcast_confirm_req (block_a);
+		}
+	}
+}
+
+void rai::bootstrap_attempt::resolve_forks ()
+{
+	std::unique_lock<std::mutex> lock (mutex);
+	if (unresolved_forks.size () > 0)
+	{
+		BOOST_LOG (node->log) << boost::str (boost::format ("%1% unresolved forks while bootstrapping") % unresolved_forks.size ());
```

Comment:
Does `BOOST_LOG` ever result in I/O?  Say, when buffers are exhausted or special characters are present in the output?